README.md
# bird-oid

[![npm version](https://img.shields.io/npm/v/bird-oid)](https://www.npmjs.com/package/bird-oid)
[![stability-stable](https://img.shields.io/badge/stability-stable-green.svg)](https://www.npmjs.com/package/bird-oid)
[![npm minzipped size](https://img.shields.io/bundlephobia/minzip/bird-oid)](https://bundlephobia.com/package/bird-oid)
[![dependencies](https://img.shields.io/librariesio/release/npm/bird-oid)](https://github.com/dmnsgn/bird-oid/blob/main/package.json)
[![types](https://img.shields.io/npm/types/bird-oid)](https://github.com/microsoft/TypeScript)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-fa6673.svg)](https://conventionalcommits.org)
[![styled with prettier](https://img.shields.io/badge/styled_with-Prettier-f8bc45.svg?logo=prettier)](https://github.com/prettier/prettier)
[![linted with eslint](https://img.shields.io/badge/linted_with-ES_Lint-4B32C3.svg?logo=eslint)](https://github.com/eslint/eslint)
[![license](https://img.shields.io/github/license/dmnsgn/bird-oid)](https://github.com/dmnsgn/bird-oid/blob/main/LICENSE.md)

A 3D boid system with accompanying emergent behaviors. Implementation mostly based on Craig Reynolds paper Steering Behaviors For Autonomous Characters.

[![paypal](https://img.shields.io/badge/donate-paypal-informational?logo=paypal)](https://paypal.me/dmnsgn)
[![coinbase](https://img.shields.io/badge/donate-coinbase-informational?logo=coinbase)](https://commerce.coinbase.com/checkout/56cbdf28-e323-48d8-9c98-7019e72c97f3)
[![twitter](https://img.shields.io/twitter/follow/dmnsgn?style=social)](https://twitter.com/dmnsgn)

![](https://raw.githubusercontent.com/dmnsgn/bird-oid/main/screenshot.gif)

## Installation

```bash
npm install bird-oid
```

## Usage

```js
import { System as BoidSystem, behaviors } from "bird-oid";

const BOIDS_COUNT = 100;
const WANDER_SPEED = 2;

const system = new BoidSystem({
  maxSpeed: 0.3,
  maxForce: 0.4,
});

const boidBehaviors = [
  {
    fn: behaviors.wander,
    options: {
      distance: system.scale * 0.04,
      radius: system.scale * 0.3,
      theta: 0,
      phi: 0.3,
    },
  },
  {
    enabled: true,
    fn: behaviors.boundsConstrain,
    scale: 2,
  },
  {
    enabled: false,
    fn: behaviors.boundsWrapConstrain,
  },
];

for (let i = 0; i < BOIDS_COUNT; i++) {
  const boid = new Boid(boidBehaviors);

  // Position in the bounds
  boid.position = system.getRandomPosition();

  // Add initial velocity
  boid.velocity = [
    Math.random() * 0.5 * system.maxSpeed,
    Math.random() * 0.5 * system.maxSpeed,
    Math.random() * 0.5 * system.maxSpeed,
  ];
  system.addBoid(boid);
}

const frame = () => {
  const dt = Math.min(0.1, myClock.getDelta());

  // Wander requires angle update
  boidBehaviors[0].options.theta +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;
  boidBehaviors[0].options.phi +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;

  system.update(dt);

  requestAnimationFrame(frame);
};

requestAnimationFrame(() => {
  frame();
});
```

## API

<!-- api-start -->

## Modules

<dl>
<dt><a href="#module_behaviors">behaviors</a></dt>
<dd><p>Steering based on <a href="https://www.red3d.com/cwr/steer/gdc99/">https://www.red3d.com/cwr/steer/gdc99/</a>
Use these function as <code>fn</code> property of the <code>BehaviorObject</code> passed to a <code>Boid</code>.</p>
</dd>
</dl>

## Classes

<dl>
<dt><a href="#Boid">Boid</a></dt>
<dd><p>A data structure for a single Boid.</p>
</dd>
<dt><a href="#Obstacle">Obstacle</a></dt>
<dd><p>Data structure used for obstacle avoidance behavior.</p>
</dd>
<dt><a href="#Path">Path</a></dt>
<dd><p>Data structure used for path following behavior.</p>
</dd>
<dt><a href="#System">System</a></dt>
<dd><p>Handle common boids update and global state.</p>
</dd>
</dl>

## Typedefs

<dl>
<dt><a href="#SystemOptions">SystemOptions</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#Radians">Radians</a> : <code>number</code></dt>
<dd></dd>
<dt><a href="#BehaviorObject">BehaviorObject</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#ApplyBehaviorObject">ApplyBehaviorObject</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#BehaviorOptions">BehaviorOptions</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#BasicBehaviorOptions">BasicBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#ExtendedBasicBehaviorOptions">ExtendedBasicBehaviorOptions</a> : <code><a href="#BasicBehaviorOptions">BasicBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#BasicWithRadiusBehaviorOptions">BasicWithRadiusBehaviorOptions</a> : <code><a href="#BasicBehaviorOptions">BasicBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#ObstacleBehaviorOptions">ObstacleBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#WanderBehaviorOptions">WanderBehaviorOptions</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#FollowPathBehaviorOptions">FollowPathBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#FollowFlowFieldSimpleBehaviorOptions">FollowFlowFieldSimpleBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#FollowFlowFieldBehaviorOptions">FollowFlowFieldBehaviorOptions</a> : <code><a href="#FollowFlowFieldSimpleBehaviorOptions">FollowFlowFieldSimpleBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#FollowLeaderSimpleBehaviorOptions">FollowLeaderSimpleBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#FollowLeaderBehaviorOptions">FollowLeaderBehaviorOptions</a> : <code><a href="#FollowLeaderSimpleBehaviorOptions">FollowLeaderSimpleBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#GroupsBehaviorOptions">GroupsBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#ConstraintBehaviorOptions">ConstraintBehaviorOptions</a> : <code><a href="#BehaviorOptions">BehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#BoundsConstraintBehaviorOptions">BoundsConstraintBehaviorOptions</a> : <code><a href="#ConstraintBehaviorOptions">ConstraintBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#SphereConstraintBehaviorOptions">SphereConstraintBehaviorOptions</a> : <code><a href="#ConstraintBehaviorOptions">ConstraintBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#WrapConstraintBehaviorOptions">WrapConstraintBehaviorOptions</a> : <code>Object</code></dt>
<dd></dd>
<dt><a href="#BoundsWrapConstraintBehaviorOptions">BoundsWrapConstraintBehaviorOptions</a> : <code><a href="#WrapConstraintBehaviorOptions">WrapConstraintBehaviorOptions</a></code></dt>
<dd></dd>
<dt><a href="#SphereWrapConstraintBehaviorOptions">SphereWrapConstraintBehaviorOptions</a> : <code><a href="#WrapConstraintBehaviorOptions">WrapConstraintBehaviorOptions</a></code></dt>
<dd></dd>
</dl>

<a name="module_behaviors"></a>

## behaviors

Steering based on https://www.red3d.com/cwr/steer/gdc99/
Use these function as `fn` property of the `BehaviorObject` passed to a `Boid`.

- [behaviors](#module_behaviors)
  - [.seek(options)](#module_behaviors.seek) ⇒ <code>module:gl-matrix~vec3</code>
  - [.flee(options)](#module_behaviors.flee) ⇒ <code>module:gl-matrix~vec3</code>
  - [.pursue(options)](#module_behaviors.pursue) ⇒ <code>module:gl-matrix~vec3</code>
  - [.evade(options)](#module_behaviors.evade) ⇒ <code>module:gl-matrix~vec3</code>
  - [.arrive(options)](#module_behaviors.arrive) ⇒ <code>module:gl-matrix~vec3</code>
  - [.avoidObstacles(options)](#module_behaviors.avoidObstacles) ⇒ <code>module:gl-matrix~vec3</code>
  - [.wander(options)](#module_behaviors.wander) ⇒ <code>module:gl-matrix~vec3</code>
  - [.followPath(options)](#module_behaviors.followPath) ⇒ <code>module:gl-matrix~vec3</code>
  - [.followFlowFieldSimple(options)](#module_behaviors.followFlowFieldSimple) ⇒ <code>module:gl-matrix~vec3</code>
  - [.followFlowField(options)](#module_behaviors.followFlowField) ⇒ <code>module:gl-matrix~vec3</code>
  - [.followLeaderSimple(options)](#module_behaviors.followLeaderSimple) ⇒ <code>module:gl-matrix~vec3</code>
  - [.followLeader(options)](#module_behaviors.followLeader) ⇒ <code>module:gl-matrix~vec3</code>
  - [.separate(options)](#module_behaviors.separate) ⇒ <code>module:gl-matrix~vec3</code>
  - [.cohere(options)](#module_behaviors.cohere) ⇒ <code>module:gl-matrix~vec3</code>
  - [.align(options)](#module_behaviors.align) ⇒ <code>module:gl-matrix~vec3</code>
  - [.flock(options)](#module_behaviors.flock) ⇒ <code>module:gl-matrix~vec3</code>
  - [.boundsConstrain(options)](#module_behaviors.boundsConstrain) ⇒ <code>module:gl-matrix~vec3</code>
  - [.sphereConstrain(options)](#module_behaviors.sphereConstrain) ⇒ <code>module:gl-matrix~vec3</code>
  - [.boundsWrapConstrain(options)](#module_behaviors.boundsWrapConstrain) ⇒ <code>module:gl-matrix~vec3</code>
  - [.sphereWrapConstrain(options)](#module_behaviors.sphereWrapConstrain) ⇒ <code>module:gl-matrix~vec3</code>

<a name="module_behaviors.seek"></a>

### behaviors.seek(options) ⇒ <code>module:gl-matrix~vec3</code>

"Seek (or pursuit of a static target) acts to steer the character towards a specified position in global space."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                       |
| ------- | ---------------------------------------------------------- |
| options | [<code>BasicBehaviorOptions</code>](#BasicBehaviorOptions) |

<a name="module_behaviors.flee"></a>

### behaviors.flee(options) ⇒ <code>module:gl-matrix~vec3</code>

"Flee is simply the inverse of seek and acts to steer the character so that its velocity is radially aligned away from the target. The desired velocity points in the opposite direction."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                       |
| ------- | ---------------------------------------------------------- |
| options | [<code>BasicBehaviorOptions</code>](#BasicBehaviorOptions) |

<a name="module_behaviors.pursue"></a>

### behaviors.pursue(options) ⇒ <code>module:gl-matrix~vec3</code>

"Pursuit is similar to seek except that the quarry (target) is another moving character. [...] The position of a character T units of time in the future (assuming it does not maneuver) can be obtained by scaling its velocity by T and adding that offset to its current position. Steering for pursuit is then simply the result of applying the seek steering behavior to the predicted target location."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                       |
| ------- | -------------------------------------------------------------------------- |
| options | [<code>ExtendedBasicBehaviorOptions</code>](#ExtendedBasicBehaviorOptions) |

<a name="module_behaviors.evade"></a>

### behaviors.evade(options) ⇒ <code>module:gl-matrix~vec3</code>

"Evasion is analogous to pursuit, except that flee is used to steer away from the predicted future position of the target character."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                       |
| ------- | -------------------------------------------------------------------------- |
| options | [<code>ExtendedBasicBehaviorOptions</code>](#ExtendedBasicBehaviorOptions) |

<a name="module_behaviors.arrive"></a>

### behaviors.arrive(options) ⇒ <code>module:gl-matrix~vec3</code>

"Arrival behavior is identical to seek while the character is far from its target. But instead of moving through the target at full speed, this behavior causes the character to slow down as it approaches the target, eventually slowing to a stop coincident with the target"

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                           |
| ------- | ------------------------------------------------------------------------------ |
| options | [<code>BasicWithRadiusBehaviorOptions</code>](#BasicWithRadiusBehaviorOptions) |

<a name="module_behaviors.avoidObstacles"></a>

### behaviors.avoidObstacles(options) ⇒ <code>module:gl-matrix~vec3</code>

"The implementation of obstacle avoidance behavior described here will make a simplifying assumption that both the character and obstacle can be reasonably approximated as spheres, although the basic concept can be easily extend to more precise shape models. [...] It is convenient to consider the geometrical situation from the character’s local coordinate system. The goal of the behavior is to keep an imaginary cylinder of free space in front of the character. The cylinder lies along the character’s forward axis, has a diameter equal to the character’s bounding sphere, and extends from the character’s center for a distance based on the character’s speed and agility. An obstacle further than this distance away is not an immediate threat. The obstacle avoidance behavior considers each obstacle in turn (perhaps using a spatial portioning scheme to cull out distance obstacles) and determines if they intersect with the cylinder. By localizing the center of each spherical obstacle, the test for non-intersection with the cylinder is very fast. The local obstacle center is projected onto the side-up plane (by setting its forward coordinate to zero) if the 2D distance from that point to the local origin is greater than the sum of the radii of the obstacle and the character, then there is no potential collision. Similarly obstacles which are fully behind the character, or fully ahead of the cylinder, can be quickly rejected. For any remaining obstacles a line-sphere intersection calculation is performed. The obstacle which intersects the forward axis nearest the character is selected as the “most threatening.” Steering to avoid this obstacle is computed by negating the (lateral) side-up projection of the obstacle’s center.

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                             |
| ------- | ---------------------------------------------------------------- |
| options | [<code>ObstacleBehaviorOptions</code>](#ObstacleBehaviorOptions) |

<a name="module_behaviors.wander"></a>

### behaviors.wander(options) ⇒ <code>module:gl-matrix~vec3</code>

"The steering force takes a “random walk” from one direction to another. This idea [...] is to constrain the steering force to the surface of a sphere located slightly ahead of the character. To produce the steering force for the next frame: a random displacement is added to the previous value, and the sum is constrained again to the sphere’s surface. The sphere’s radius [...] determines the maximum wandering “strength” and the magnitude of the random displacement [...] determines the wander “rate.”"
Note: Don't forget to update theta and phi angles: angle += Math.random() _ WANDER_SPEED - WANDER_SPEED _ 0.5;

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                         |
| ------- | ------------------------------------------------------------ |
| options | [<code>WanderBehaviorOptions</code>](#WanderBehaviorOptions) |

<a name="module_behaviors.followPath"></a>

### behaviors.followPath(options) ⇒ <code>module:gl-matrix~vec3</code>

"Path following behavior enables a character to steer along a predetermined path, such as a roadway, corridor or tunnel."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                 |
| ------- | -------------------------------------------------------------------- |
| options | [<code>FollowPathBehaviorOptions</code>](#FollowPathBehaviorOptions) |

<a name="module_behaviors.followFlowFieldSimple"></a>

### behaviors.followFlowFieldSimple(options) ⇒ <code>module:gl-matrix~vec3</code>

"In flow field following behavior the character steers to align its motion with the local tangent of a flow field (also known as a force field or a vector field). [...] The future position of a character is estimated and the flow field is sampled at that location. This flow direction [...] is the “desired velocity” and the steering direction (vector S) is simply the difference between the current velocity (vector V) and the desired velocity."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                                       |
| ------- | ------------------------------------------------------------------------------------------ |
| options | [<code>FollowFlowFieldSimpleBehaviorOptions</code>](#FollowFlowFieldSimpleBehaviorOptions) |

<a name="module_behaviors.followFlowField"></a>

### behaviors.followFlowField(options) ⇒ <code>module:gl-matrix~vec3</code>

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                           |
| ------- | ------------------------------------------------------------------------------ |
| options | [<code>FollowFlowFieldBehaviorOptions</code>](#FollowFlowFieldBehaviorOptions) |

<a name="module_behaviors.followLeaderSimple"></a>

### behaviors.followLeaderSimple(options) ⇒ <code>module:gl-matrix~vec3</code>

"Leader following behavior causes one or more character to follow another moving character designated as the leader. Generally the followers want to stay near the leader, without crowding the leader, and taking care to stay out of the leader’s way (in case they happen to find them selves in front of the leader). In addition, if there is more than one follower, they want to avoid bumping each other. The implementation of leader following relies on arrival behavior (see above) a desire to move towards a point, slowing as it draws near. The arrival target is a point offset slightly behind the leader. (The offset distance might optionally increases with speed.) If a follower finds itself in a rectangular region in front of the leader, it will steer laterally away from the leader’s path before resuming arrival behavior. In addition the followers use separation behavior to prevent crowding each other."

**Kind**: static method of [<code>behaviors</code>](#module_behaviors)

| Param   | Type                                                                                 |
| ------- | ------------------------------------------------------------------------------------ |
| options | [<code>FollowLeaderSimpleBehaviorOptions</code>](#FollowLeaderSimpleBehaviorOptions) |

<a name="module_behaviors.followLeader"></a>

### behaviors.followLeader(options) ⇒ <code>module:gl-matrix~vec3</code>

Notes: next behaviour should be a separation


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.2.1](https://github.com/dmnsgn/bird-oid/compare/v0.2.0...v0.2.1) (2022-04-26)


demo.js
import SimplexNoise from "simplex-noise";

import {
  Clock,
  WebGLRenderer,
  Scene,
  PerspectiveCamera,
  Object3D,
  Vector3,
  Quaternion,
  InstancedBufferAttribute,
  CylinderBufferGeometry,
  BoxBufferGeometry,
  SphereGeometry,
  SphereBufferGeometry,
  MeshNormalMaterial,
  MeshBasicMaterial,
  InstancedMesh,
  Mesh,
  BoxHelper,
  AxesHelper,
  ArrowHelper,
  VertexColors,
  Color,
  TubeBufferGeometry,
  CurvePath,
  LineCurve3,
  DoubleSide
} from "three";

import { OrbitControls } from "three/examples/jsm/controls/OrbitControls";

import VectorField from "vector-field";

import * as dat from "dat.gui";

import {
  System as BoidSystem,
  Boid,
  Obstacle,
  behaviors,
  Path
} from "./index.js";

const system = new BoidSystem({
  maxSpeed: 0.9,
  maxForce: 0.8
});
const systemHalfScale = system.scale * 0.5;
const clock = new Clock();

const ENTITIES = {
  targets: [],
  boids: [],
  obstacles: []
};

// Add target boid
const WANDER_SPEED = 2;
const targetWanderBehavior = {
  fn: behaviors.wander,
  options: {
    distance: system.scale * 0.04,
    radius: system.scale * 0.3,
    theta: 0,
    phi: 0.3
  }
};
const targetBoid = new Boid([
  targetWanderBehavior,
  { fn: behaviors.boundsConstrain, scale: 1 }
]);
targetBoid.position = system.getRandomPosition();
system.addBoid(targetBoid);
ENTITIES.targets.push(targetBoid);

// FLOW FIELD
const tempDirection = new Vector3();
const tempColor = new Color();

const simplex = new SimplexNoise(Math.random);
function noiseFn([x, y, z], [stepX, stepY, stepZ]) {
  const scale = 0.5;

  let n =
    simplex.noise4D(x * scale, y * scale, z * scale, 0.05 * clock.elapsedTime) *
    Math.PI *
    2;
  const theta = n;
  const phi = n;

  const direction = [
    Math.sin(theta) * Math.sin(phi),
    Math.cos(theta),
    Math.sin(theta) * Math.cos(phi)
  ];

  const mesh = this.field[stepX][stepY][stepZ].mesh;

  if (mesh) {
    mesh.setDirection(tempDirection.fromArray(direction));
    mesh.setColor(tempColor.fromArray([...direction, 1]));
  }

  return direction;
}
const flowField = new VectorField(noiseFn, 12, system.scale);
flowField.update();

const pathPointsCount = 5;
const path = new Path(
  Array.from({ length: pathPointsCount }, (_, i) => {
    const offset = Math.random() * system.scale * 0.25;
    const radius = systemHalfScale * 0.75;

    const theta = (0.8 * i) / (pathPointsCount - 1);
    return [
      radius * Math.cos(theta * Math.PI * 2),
      offset,
      radius * Math.sin(theta * Math.PI * 2)
    ];
  }),
  system.scale * 0.03
);

// Config
const CONFIG = {
  cameraFollowTarget: false,
  entities: {
    boids: {
      enabled: true,
      count: 200
    },
    targets: {
      enabled: true,
      count: 1
    },
    obstacles: {
      enabled: false,
      count: 30
    }
  },
  behaviors: [
    {
      enabled: false,
      fn: behaviors.seek,
      scale: 1,
      options: {
        target: [0, 0, 0]
      }
    },
    {
      enabled: false,
      fn: behaviors.flee,
      scale: 1,
      options: {
        target: [0, 0, 0]
      }
    },
    {
      enabled: false,
      fn: behaviors.pursue,
      scale: 1,
      options: {
        target: targetBoid.position,
        targetVelocity: targetBoid.velocity
      }
    },
    {
      enabled: true,
      fn: behaviors.evade,
      scale: 1,
      options: {
        target: targetBoid.position,
        targetVelocity: targetBoid.velocity
      }
    },
    {
      enabled: false,
      fn: behaviors.arrive,
      scale: 1,
      options: {
        radius: system.scale * 1
      }
    },
    {
      enabled: true,
      fn: behaviors.avoidObstacles,
      scale: 4,
      options: {
        maxAvoidForce: system.scale * 2,
        obstacles: ENTITIES.obstacles
      }
    },
    {
      enabled: false,
      fn: behaviors.wander,
      scale: 1,
      options: {
        distance: system.scale * 0.04,
        radius: system.scale * 0.3,
        theta: 0,
        phi: 0.2
      }
    },
    {
      enabled: true,
      fn: behaviors.followPath,
      scale: 1,
      options: {
        path,
        maxSpeed: 0.3,
        maxForce: 0.4
      }
    },
    {
      enabled: false,
      fn: behaviors.followFlowFieldSimple,
      scale: 1,
      options: {
        flowField
      }
    },
    {
      enabled: false,
      fn: behaviors.followFlowField,
      scale: 1,
      options: {
        flowField
      }
    },
    {
      enabled: true,
      fn: behaviors.separate,
      scale: 1,
      options: {
        maxDistance: system.scale * 0.04
      }
    },
    {
      enabled: false,
      fn: behaviors.cohere,
      scale: 1,
      options: {
        maxDistance: system.scale * 0.04
      }
    },
    {
      enabled: false,
      fn: behaviors.align,
      scale: 1,
      options: {
        maxDistance: system.scale * 0.04
      }
    },
    {
      enabled: false,
      fn: behaviors.flock,
      scale: 1,
      options: {
        maxDistance: system.scale * 0.09
      }
    },
    {
      enabled: false,
      fn: behaviors.followLeaderSimple,
      scale: 1,
      options: {
        leader: targetBoid,
        distance: system.scale * 0.04,
        radius: system.scale * 0.3
      }
    },
    {
      enabled: false,
      fn: behaviors.followLeader,
      scale: 1,
      options: {
        leader: targetBoid,
        distance: system.scale * 0.04,
        radius: system.scale * 0.3
      }
    },
    {
      enabled: false,
      fn: behaviors.boundsConstrain,
      scale: 2
    },
    {
      enabled: true,
      fn: behaviors.sphereConstrain,
      scale: 2,
      options: {
        radius: systemHalfScale
      }
    },
    {
      enabled: false,
      fn: behaviors.boundsWrapConstrain
    },
    {
      enabled: false,
      fn: behaviors.sphereWrapConstrain,
      options: {
        radius: systemHalfScale
      }
    }
  ]
};

// GUI
const gui = new dat.GUI();
const systemFolder = gui.addFolder("System");
systemFolder.add(system, "scale", 0, 10);
systemFolder.add(system, "maxSpeed", 0, 10);
systemFolder.add(system, "maxForce", 0, 10);
systemFolder.add(CONFIG, "cameraFollowTarget");
systemFolder.add(CONFIG.entities.boids, "enabled").name("Boids");
systemFolder.add(CONFIG.entities.targets, "enabled").name("Targets");
systemFolder.add(CONFIG.entities.obstacles, "enabled").name("Obstacles");

for (let i = 0; i < CONFIG.behaviors.length; i++) {
  const behavior = CONFIG.behaviors[i];
  const name = behavior.fn.name;
  const behaviorFolder = gui.addFolder(name);
  // behaviorFolder.open();
  behaviorFolder.add(behavior, "enabled");
  if (behavior.scale) behaviorFolder.add(behavior, "scale");
}

// SCENE
const tempObject = new Object3D();
const tempQuaternion = new Quaternion();
const renderer = new WebGLRenderer();
renderer.setPixelRatio(window.devicePixelRatio);
renderer.setSize(window.innerWidth, window.innerHeight);

const scene = new Scene();
const camera = new PerspectiveCamera(
  75,
  window.innerWidth / window.innerHeight,
  0.1,
  1000
);
Object.assign(camera.position, {
  x: 0.5,
  y: 0.5,
  z: 1.5
});
const controls = new OrbitControls(camera, renderer.domElement);
Object.assign(controls, {
  enableDamping: true,
  dampingFactor: 0.05,
  screenSpacePanning: false,
  maxDistance: 50
});

// ENTITIES
const getBoidGeometry = () => {
  const geometry = new CylinderBufferGeometry(0, 0.01, 0.03, 3, 1);
  geometry.rotateX(Math.PI / 2);
  return geometry;
};

const boidsMesh = new InstancedMesh(
  getBoidGeometry(),
  new MeshNormalMaterial(),
  CONFIG.entities.boids.count
);
scene.add(boidsMesh);

const targetMesh = new InstancedMesh(
  getBoidGeometry(),
  new MeshBasicMaterial({ color: 0x00ff00 }),
  CONFIG.entities.targets.count
);
scene.add(targetMesh);

const obstacleGeometry = new SphereBufferGeometry(1);
const colorBuffer = new Float32Array(CONFIG.entities.obstacles.count * 3);
for (let i = 0; i < CONFIG.entities.obstacles.count; i++) {
  colorBuffer[i * 3 + 0] = 1;
  colorBuffer[i * 3 + 1] = Math.random();
  colorBuffer[i * 3 + 2] = 0;
}
obstacleGeometry.setAttribute(
  "color",
  new InstancedBufferAttribute(colorBuffer, 3)
);

const obstacleMaterial = new MeshBasicMaterial();
obstacleMaterial.vertexColors = VertexColors;

const obstacleMesh = new InstancedMesh(
  obstacleGeometry,
  obstacleMaterial,
  CONFIG.entities.obstacles.count
);
scene.add(obstacleMesh);

// INIT SYSTEM
for (let i = 0; i < CONFIG.entities.boids.count; i++) {
  const boid = new Boid(CONFIG.behaviors);
  boid.position = system.getRandomPosition();
  boid.velocity = [
    Math.random() * 0.5 * system.maxSpeed,
    Math.random() * 0.5 * system.maxSpeed,
    Math.random() * 0.5 * system.maxSpeed
  ];
  system.addBoid(boid);
  ENTITIES.boids.push(boid);
}

for (let i = 0; i < CONFIG.entities.obstacles.count; i++) {
  const radius = system.scale * 0.025 + Math.random() * system.scale * 0.025;
  const obstacle = new Obstacle(system.getRandomPosition(), radius);
  ENTITIES.obstacles.push(obstacle);
}

// DEBUG
const debugMeshes = new Object3D();
const debugFolder = gui.addFolder("Debug");
debugFolder.add(debugMeshes, "visible").name("all");

const debugAxes = new AxesHelper(1);
debugFolder.add(debugAxes, "visible").name("axes");
debugMeshes.add(debugAxes);

const debugBounds = new BoxHelper(
  new Mesh(
    new BoxBufferGeometry(system.scale, system.scale, system.scale),
    new MeshBasicMaterial()
  ),
  0xffffff
);
debugBounds.visible = false;
debugFolder.add(debugBounds, "visible").name("bounds");
debugMeshes.add(debugBounds);

const debugSphereBounds = new Mesh(
  new SphereGeometry(systemHalfScale),
  new MeshBasicMaterial({
    color: 0xffffff,
    wireframe: true,
    transparent: true,
    opacity: 0.2
  })
);
// debugBounds.visible = false;
debugFolder.add(debugSphereBounds, "visible").name("sphere bounds");
debugMeshes.add(debugSphereBounds);

const curvePath = new CurvePath();
for (let i = 1; i < path.points.length; i++) {
  const prev = path.points[i - 1];
  const current = path.points[i];
  curvePath.add(
    new LineCurve3(
      new Vector3().fromArray(prev),
      new Vector3().fromArray(current)
    )
  );
}

const debugPath = new Mesh(
  new TubeBufferGeometry(curvePath, path.points.length, path.radius, 8, false),
  new MeshBasicMaterial({
    color: 0xffffff,
    wireframe: true,
    side: DoubleSide
  })
);
debugPath.visible = false;
debugFolder.add(debugPath, "visible").name("path");
debugMeshes.add(debugPath);

const debugFlowField = new Object3D();
for (let x = 0; x < flowField.steps[0]; x++) {
  for (let y = 0; y < flowField.steps[1]; y++) {
    for (let z = 0; z < flowField.steps[2]; z++) {
      const mesh = new ArrowHelper(
        tempDirection.fromArray(flowField.field[x][y][z].direction),
        new Vector3()
          .fromArray(flowField.field[x][y][z].position)
          .sub(new Vector3(systemHalfScale, systemHalfScale, systemHalfScale)),
        system.scale * 0.05,
        0xffffff
      );
      debugFlowField.add(mesh);
      flowField.field[x][y][z].mesh = mesh;
    }
  }
}
debugFlowField.visible = false;
debugFolder.add(debugFlowField, "visible").name("field");
debugMeshes.add(debugFlowField);
scene.add(debugMeshes);

// LOOP
const frame = () => {
  const dt = Math.min(0.1, clock.getDelta());

  // System
  system.update(dt);

  // Flow field
  flowField.update(); // See noiseFn for debug meshes update

  // Obstacles
  // TODO: CONFIG.obstaclesMoving
  obstacleMesh.visible = CONFIG.entities.obstacles.enabled;
  if (obstacleMesh.visible) {
    for (let i = 0; i < ENTITIES.obstacles.length; i++) {
      const obstacle = ENTITIES.obstacles[i];
      tempObject.position.set(...obstacle.position);
      tempObject.rotation.set(0, 0, 0);
      tempObject.scale.set(obstacle.radius, obstacle.radius, obstacle.radius);
      tempObject.updateMatrix();
      obstacleMesh.setMatrixAt(i, tempObject.matrix);
    }
    obstacleMesh.instanceMatrix.needsUpdate = true;
  }

  tempObject.scale.set(1, 1, 1);

  // Target
  const configWanderBehavior = CONFIG.behaviors.find(
    behavior => behavior.fn.name === "wander"
  );
  configWanderBehavior.options.theta +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;
  configWanderBehavior.options.phi +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;
  targetWanderBehavior.options.theta +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;
  targetWanderBehavior.options.phi +=
    Math.random() * WANDER_SPEED - WANDER_SPEED * 0.5;

  targetMesh.visible = CONFIG.entities.targets.enabled;
  if (targetMesh.visible) {
    for (let i = 0; i < ENTITIES.targets.length; i++) {
      const target = ENTITIES.targets[i];
      tempObject.position.set(...target.position);
      tempObject.rotation.setFromQuaternion(
        tempQuaternion.setFromUnitVectors(
          new Vector3(0, 0, 1),
          new Vector3(...target.velocity).normalize()
        )
      );
      tempObject.updateMatrix();
      targetMesh.setMatrixAt(i, tempObject.matrix);
    }
    targetMesh.instanceMatrix.needsUpdate = true;
  }

  // Boids
  boidsMesh.visible = CONFIG.entities.boids.enabled;
  if (boidsMesh.visible) {
    for (let i = 0; i < ENTITIES.boids.length; i++) {
      const boid = ENTITIES.boids[i];
      tempObject.position.set(...boid.position);
      tempObject.rotation.setFromQuaternion(
        tempQuaternion.setFromUnitVectors(
          new Vector3(0, 0, 1),
          new Vector3(...boid.velocity).normalize()
        )
      );
      tempObject.updateMatrix();
      boidsMesh.setMatrixAt(i, tempObject.matrix);
    }
    boidsMesh.instanceMatrix.needsUpdate = true;
  }

  if (CONFIG.cameraFollowTarget) {
    camera.lookAt(new Vector3().fromArray(targetBoid.position));
  }

  renderer.render(scene, camera);
  controls.update();

  requestAnimationFrame(frame);
};

// Kick off
requestAnimationFrame(() => {
  document.body.appendChild(renderer.domElement);
  clock.start();
  frame();
});

// Events
window.addEventListener("resize", () => {
  camera.aspect = window.innerWidth / window.innerHeight;
  camera.updateProjectionMatrix();
  renderer.setSize(window.innerWidth, window.innerHeight);
});
